File: app/src/main/java/org/schabi/newpipe/player/notification/NotificationPlayerUi.java
package org.schabi.newpipe.player.notification;

import static org.schabi.newpipe.player.helper.PlayerHelper.MinimizeMode.MINIMIZE_ON_EXIT_MODE_NONE;
import static org.schabi.newpipe.player.notification.NotificationConstants.ACTION_RECREATE_NOTIFICATION;

import android.content.Intent;
import android.graphics.Bitmap;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.exoplayer2.Player.RepeatMode;

import org.schabi.newpipe.extractor.stream.StreamInfo;
import org.schabi.newpipe.player.Player;
import org.schabi.newpipe.player.helper.PlayerHelper;
import org.schabi.newpipe.player.ui.PlayerUi;

public final class NotificationPlayerUi extends PlayerUi {
    private final NotificationUtil notificationUtil;

    public NotificationPlayerUi(@NonNull final Player player) {
        super(player);
        notificationUtil = new NotificationUtil(player);
    }

    @Override
    public void destroy() {
        super.destroy();
        notificationUtil.cancelNotificationAndStopForeground();
    }

    @Override
    public void onThumbnailLoaded(@Nullable final Bitmap bitmap) {
        super.onThumbnailLoaded(bitmap);
        notificationUtil.updateThumbnail();
    }

    @Override
    public void onBlocked() {
        super.onBlocked();
        notificationUtil.createNotificationIfNeededAndUpdate(false);
    }

    @Override
    public void onPlaying() {
        super.onPlaying();
        notificationUtil.createNotificationIfNeededAndUpdate(false);
    }

    @Override
    public void onBuffering() {
        super.onBuffering();
        if (notificationUtil.shouldUpdateBufferingSlot()) {
            notificationUtil.createNotificationIfNeededAndUpdate(false);
        }
    }

    @Override
    public void onPaused() {
        super.onPaused();

        // Remove running notification when user does not want minimization to background or popup
        if (PlayerHelper.getMinimizeOnExitAction(context) == MINIMIZE_ON_EXIT_MODE_NONE
                && player.videoPlayerSelected()) {
            notificationUtil.cancelNotificationAndStopForeground();
        } else {
            notificationUtil.createNotificationIfNeededAndUpdate(false);
        }
    }

    @Override
    public void onPausedSeek() {
        super.onPausedSeek();
        notificationUtil.createNotificationIfNeededAndUpdate(false);
    }

    @Override
    public void onCompleted() {
        super.onCompleted();
        notificationUtil.createNotificationIfNeededAndUpdate(false);
    }

    @Override
    public void onRepeatModeChanged(@RepeatMode final int repeatMode) {
        super.onRepeatModeChanged(repeatMode);
        notificationUtil.createNotificationIfNeededAndUpdate(false);
    }

    @Override
    public void onShuffleModeEnabledChanged(final boolean shuffleModeEnabled) {
        super.onShuffleModeEnabledChanged(shuffleModeEnabled);
        notificationUtil.createNotificationIfNeededAndUpdate(false);
    }

    @Override
    public void onBroadcastReceived(final Intent intent) {
        super.onBroadcastReceived(intent);
        if (ACTION_RECREATE_NOTIFICATION.equals(intent.getAction())) {
            notificationUtil.createNotificationIfNeededAndUpdate(true);
        }
    }

    @Override
    public void onMetadataChanged(@NonNull final StreamInfo info) {
        super.onMetadataChanged(info);
        notificationUtil.createNotificationIfNeededAndUpdate(true);
    }

    @Override
    public void onPlayQueueEdited() {
        super.onPlayQueueEdited();
        notificationUtil.createNotificationIfNeededAndUpdate(false);
    }

    public void createNotificationAndStartForeground() {
        notificationUtil.createNotificationAndStartForeground();
    }
}

File: app/src/main/java/org/schabi/newpipe/player/notification/NotificationPlayerUi.kt
@file:OptIn(androidx.media3.common.util.UnstableApi::class)

package org.schabi.newpipe.player.notification

import android.content.Intent
import android.graphics.Bitmap
import androidx.media3.common.Player.RepeatMode
import org.schabi.newpipe.extractor.stream.StreamInfo
import org.schabi.newpipe.player.Player
import org.schabi.newpipe.player.helper.PlayerHelper
import org.schabi.newpipe.player.helper.PlayerHelper.MINIMIZE_ON_EXIT_MODE_NONE
import org.schabi.newpipe.player.ui.PlayerUi

class NotificationPlayerUi(player: Player) : PlayerUi(player) {
    private val notificationUtil = NotificationUtil(player)

    override fun destroy() {
        super.destroy()
        notificationUtil.cancelNotificationAndStopForeground()
    }

    override fun onThumbnailLoaded(bitmap: Bitmap?) {
        super.onThumbnailLoaded(bitmap)
        notificationUtil.updateThumbnail()
    }

    override fun onBlocked() {
        super.onBlocked()
        notificationUtil.createNotificationIfNeededAndUpdate(false)
    }

    override fun onPlaying() {
        super.onPlaying()
        notificationUtil.createNotificationIfNeededAndUpdate(false)
    }

    override fun onBuffering() {
        super.onBuffering()
        if (notificationUtil.shouldUpdateBufferingSlot()) {
            notificationUtil.createNotificationIfNeededAndUpdate(false)
        }
    }

    override fun onPaused() {
        super.onPaused()

        // Remove running notification when user does not want minimization to background or popup
        if (PlayerHelper.getMinimizeOnExitAction(context) == MINIMIZE_ON_EXIT_MODE_NONE &&
            player.videoPlayerSelected()
        ) {
            notificationUtil.cancelNotificationAndStopForeground()
        } else {
            notificationUtil.createNotificationIfNeededAndUpdate(false)
        }
    }

    override fun onPausedSeek() {
        super.onPausedSeek()
        notificationUtil.createNotificationIfNeededAndUpdate(false)
    }

    override fun onCompleted() {
        super.onCompleted()
        notificationUtil.createNotificationIfNeededAndUpdate(false)
    }

    override fun onRepeatModeChanged(@RepeatMode repeatMode: Int) {
        super.onRepeatModeChanged(repeatMode)
        notificationUtil.createNotificationIfNeededAndUpdate(false)
    }

    override fun onShuffleModeEnabledChanged(shuffleModeEnabled: Boolean) {
        super.onShuffleModeEnabledChanged(shuffleModeEnabled)
        notificationUtil.createNotificationIfNeededAndUpdate(false)
    }

    override fun onBroadcastReceived(intent: Intent) {
        super.onBroadcastReceived(intent)
        if (NotificationConstants.ACTION_RECREATE_NOTIFICATION == intent.action) {
            notificationUtil.createNotificationIfNeededAndUpdate(true)
        }
    }

    override fun onMetadataChanged(info: StreamInfo) {
        super.onMetadataChanged(info)
        notificationUtil.createNotificationIfNeededAndUpdate(true)
    }

    override fun onPlayQueueEdited() {
        super.onPlayQueueEdited()
        notificationUtil.createNotificationIfNeededAndUpdate(false)
    }

    fun createNotificationAndStartForeground() {
        notificationUtil.createNotificationAndStartForeground()
    }
}
